src/fips_check.c

```diff
@@ -0,0 +1,224 @@
+	int rc = -1;
+
+	dl = dlopen(libname, RTLD_LAZY);
+	if (dl == NULL)
+		goto done;
+
+	sym = dlsym(dl, symbolname);
+	if (sym != NULL && dladdr(sym, &info)) {
+		if (strlen(info.dli_fname) <= pathlen)
+			strcpy(path, info.dli_fname);
+		else
+			goto done;
+	}
+
+	dlclose(dl);
+	rc = 0;
+
+done:
+
+	return rc;
+}
+
+static char *make_hmac_path(const char *origpath)
+{
+	char *path, *p;
+	const char *fn;
+
+	path = malloc(sizeof(HMAC_PREFIX) + sizeof(HMAC_SUFFIX) + strlen(origpath));
+	if (path == NULL)
+		return NULL;
+
+	fn = strrchr(origpath, '/');
+	if (fn == NULL) {
+		fn = origpath;
+	} else {
+		++fn;
+	}
+
+	strncpy(path, origpath, fn - origpath);
+	p = path + (fn - origpath);
+	p = stpcpy(p, HMAC_PREFIX);
+	p = stpcpy(p, fn);
+	p = stpcpy(p, HMAC_SUFFIX);
+
+	return path;
+}
+
+static int compute_file_hmac(const char *path, void **buf, size_t *hmaclen)
+{
+	FILE *fp = NULL;
+	int rc = -1;
+	unsigned char rbuf[READ_BUFFER_LENGTH];
+	unsigned int hlen;
+	size_t len;
+	HMAC_CTX *c;
+
+	c = HMAC_CTX_new();
+	if (c == NULL)
+		return rc;
+
+	fp = fopen(path, "r");
+	if (fp == NULL)
+		goto end;
+
+	if (HMAC_Init_ex(c, hmackey, sizeof(hmackey) - 1, EVP_sha256(), NULL) <= 0)
+		goto end;
+
+	while ((len = fread(rbuf, 1, sizeof(rbuf), fp)) != 0) {
+		if (HMAC_Update(c, rbuf, len) <= 0) {
+			goto end;
+		}
+	}
+
+	len = sizeof(rbuf);
+	/* reuse rbuf for hmac */
+	if (HMAC_Final(c, rbuf, &hlen) <= 0)
+		goto end;
+
+	*buf = malloc(hlen);
+	if (*buf == NULL)
+		goto end;
+
+	*hmaclen = hlen;
+
+	memcpy(*buf, rbuf, hlen);
+
+	rc = 0;
+
+end:
+
+	HMAC_CTX_free(c);
+	if (fp)
+		fclose(fp);
+
+	return rc;
+}
+
+/**
+ * Performs the FIPS check.
+ *
+ * @return  1 if check succeeded
+ *          0 otherwise
+ */
+static int FIPSCHECK_verify(const char *path)
+{
+	int rc = 0;
+	FILE *fp;
+	unsigned char *hmac_buf = NULL;
+	long hmaclen;
+	char *hmacpath, *p;
+	char *hmac_str = NULL;
+	size_t n, buflen;
+	void *buf = NULL;
+
+	hmacpath = make_hmac_path(path);
+	if (hmacpath == NULL)
+		return 0;
+
+	fp = fopen(hmacpath, "r");
+	if (fp == NULL) {
+		rc = 1;
+		goto end;
+	}
+
+	if (getline(&hmac_str, &n, fp) <= 0)
+		goto end;
+
+	if ((p = strchr(hmac_str, '\n')) != NULL)
+		*p = '\0';
+
+	hmac_buf = OPENSSL_hexstr2buf(hmac_str, &hmaclen);
+
+	if (compute_file_hmac(path, &buf, &buflen) != 0)
+		goto end;
+
+	if (memcmp(buf, hmac_buf, hmaclen) != 0)
+		goto end;
+
+	rc = 1;
+
+end:
+
+	if (buf)
+		free(buf);
+	if (hmac_str)
+		free(hmac_str);
+	if (hmacpath)
+		free(hmacpath);
+
+	OPENSSL_free(hmac_buf);
+
+	if (fp)
+		fclose(fp);
+
+	return rc;
+}
+
+static const char msg1[] = "Libica FIPS library integrity check failed. Cannot determine library path.\n";
+static const char msg2[] = "Libica FIPS library integrity check failed. Module %s probably corrupted.\n";
+static const char msg3[] = "Libica FIPS library integrity check passed.\n";
+
+void fips_lib_integrity_check(void)
+{
+	int rc;
+	char path[PATH_MAX];
+	char *libname = "libica.so";
```

Comment:
Do we need this a little bit more high sophisticated ?
For example:
- check if libica.so is a softlink
- if yes, find the source of the softlink
- again check for softlink until finally the real so file is found.
- Now use the real so file name and the derived hmac file for the checks.
The advantage would be to able to hold more than one libica.so.xxx with all the related libica.so.xxx.hmac files.

Reply:
Good idea! But let's add this in a separate step.

Reply:
Don't we have versioned so names nowadays? E.g. libica.so.3...
I would make the so name a define, so that it can be easily changed, and possibly patched by the distributor when they build versioned so names.

Reply:
Yes, but the get_library_path routine already returns the correct path including major number, e.g. "libica.so.3". And the make_hmac_path routine produces the correct corresponding pathname of the .hmac file, e.g. /usr/lib/.libica.so.3.hmac. So there's no need to specify the major number somewhere in the code.